### ISSUE TITLE
Animate the hamburger icon when selecting a View

### ISSUE THREAD

**Issue by zaclimon** [mikepenz/MaterialDrawer#739]
Hey,

I don't know if this is an issue or not but I'm searching to replicate the hamburger/back arrow animation like in the Material Design guidelines shown here:

https://material-design.storage.googleapis.com/publish/material_v_4/material_ext_publish/0B3T7oTWa3HiFbFRfT196SWRyS2s/animation_delightfuldetails_wellcrafted.webm

I know it is possible to do it when opening/closing the drawer but what I would like to do is like as shown, selecting a view and then animate the arrow.

I've done some research and I see that you have to enable the "spinBars" option in the styles as shown here: http://stackoverflow.com/a/26447144
That said, after checking the styles sources, it looks like this option has been disabled on purpose.

 @mikepenz do you think it is possible to implement a method that would show the hamburger icon/back arrow animated from within Material Drawer?

Thanks.

**Comment by mikepenz** [comment]
@zaclimon there is the `withActionBarDrawerToggleAnimated(true)` method you can set on the `DrawerBuilder` it will animate the arrow if the drawer is opened / closed.

**Comment by zaclimon** [comment]
@mikepenz Yeah I know about the withActionBarDrawerToggleAnimated() method, that said even though it animates the arrow when the Drawer is opened/closed, it doesn't animate when I want to transform it.

See here a screen record regarding what I meant:

https://www.youtube.com/watch?v=6Ea1CohUrvM

Using the recommended way of showing the back arrow, you can see that the arrow is animated when trying to open/close the Drawer. However, it doesn't animate when accessing the wanted section of the app.

**Comment by mikepenz** [comment]
@zaclimon yes that's the normal behavior. 

The `ActionBarDrawerToggle` is independent of the **MaterialDrawer** and this is no library specific topic. 

But to help you get into the right direction. I think you will have to get the `ActionBarDrawerToggle` from the `Drawer` object and then animate it manually like done here:

http://stackoverflow.com/a/26480439/325479

**Comment by zaclimon** [comment]
Hi,

I don't know if somebody will have the same situation eventually but I managed to fix this thanks for the link @mikepenz. I used the animation from the StackOverFlow link and I applied also a `OnDrawerNavigationListener` so I can go a step backwards when pressing the back arrow.


That said, I had an issue being that the hamburger --> back arrow animation wasn't working but showing perfectly the othjer way around. I fixed it by adding the required methods to show the back arrow by adding a ValueAnimator listener to wait until the animation is done before changing the arrow's method.

**Comment by mikepenz** [comment]
@zaclimon thanks for adding your solution here

**Comment by zaclimon** [comment]
@mikepenz Sorry to bring up this old topic again, that said, I have to nitpick over something. While you said the ActionBarDrawerToggle is independent from the Drawer, the method responsible for it to turn over is not. `(onDrawerSlide)`

At the time, I was looking to keep the toggle not animated while drawing the slider but animate it when necessary. (Simillar to the Play Store) It looks like the Drawer icon should not animate when the Drawer is sliding above it. (I may be wrong)

By checking in DrawerBuilder.java here:

https://github.com/mikepenz/MaterialDrawer/blob/develop/library/src/main/java/com/mikepenz/materialdrawer/DrawerBuilder.java#L1368

if the `.withActionBarDrawerToggleAnimated` is set to false, no matter what I set to `onDrawerSlide`, it will always return a value of 0. (And thus the drawer toggle not turning as wanted)

I tried to create a temporary ActionBarToggle that could replicate the animation. Whilst I was able to create an beginning animation, without access to the Drawer listener, The back arrow can't do anything.

I understand if you don't see this as important as it is not much of a detail.

**Comment by mikepenz** [comment]
@zaclimon the `mAnimateActionBarDrawerToggle` is only true if a toolbar was provided and the animation is activated. 

If you want to hook in the drawer slide, just add a `OnDrawerListener` to your `DrawerBuilder`, and you are fine.

**Comment by zaclimon** [comment]
@mikepenz I do provide a Toolbar when I'm creating the drawer which is why I can get the ActionBarDrawerToggle. If you don't mind, here's the code responsible for turning the arrow:

    public void showBackToolbarArrow(boolean enabled) {

        final ActionBarDrawerToggle toggle = mDrawer.getActionBarDrawerToggle();
        ValueAnimator valueAnimator;

        if (enabled) {
            valueAnimator = ValueAnimator.ofFloat(0, 1);
            mDrawer.getDrawerLayout().setDrawerLockMode(DrawerLayout.LOCK_MODE_LOCKED_CLOSED);

            valueAnimator.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    if (getSupportActionBar() != null) {
                        toggle.setDrawerIndicatorEnabled(false);
                        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
                    }
                }
            });
        } else {
            valueAnimator = ValueAnimator.ofFloat(1, 0);
            mDrawer.getDrawerLayout().setDrawerLockMode(DrawerLayout.LOCK_MODE_UNLOCKED);

            valueAnimator.addListener(new AnimatorListenerAdapter() {

                @Override
                public void onAnimationStart(Animator animation) {
                    super.onAnimationStart(animation);
                    isTweaksExiting = true;
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    isTweaksExiting = false;
                }
            });

            if (getSupportActionBar() != null) {
                getSupportActionBar().setDisplayHomeAsUpEnabled(false);
                toggle.setDrawerIndicatorEnabled(true);
            }
        }

        // Play the hamburger icon/back arrow animation based on the ValueAnimator.ofFloat.
        valueAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                float animationOffset = (float) animation.getAnimatedValue();
                toggle.onDrawerSlide(mDrawer.getDrawerLayout(), animationOffset);
            }
        });
        valueAnimator.setDuration(400);
        valueAnimator.setInterpolator(new AccelerateDecelerateInterpolator());
        valueAnimator.start();
    }


If I enable `.withActionBarDrawerToggleAnimated`, toggle.onDrawerSlide will work since the super call will return the correct offset value. Now as I said earlier, unless I somehow create a temporary ActionBarDrawerToggle, (and break the one made by DrawerBuilder) I won't be able to call toggle.onDrawerSlide if `.withActionBarDrawerToggleAnimated` is disabled. I could use the onDrawerListener but looks like it'll end up like a pretty mess.

I'll continue to investigate to see if there is a way to do this. Thanks for the help!

**Comment by mikepenz** [comment]
@zaclimon and what is against doing the `ActionBarDrawerToggle` on your own? Just don't provide the `Toolbar` create your `ActionBarDrawerToggle` get the `DrawerLayout` from the `Drawer` and you can have your custom behavior as you need it?

**Comment by zaclimon** [comment]
@mikepenz Yeah that's what I ended up doing. I thought that we couldn't put any custom `ActionBarDrawerToggle` after creating the `Drawer`. Though there is the `Drawer.setActionBarDrawerToggle` method that I didn't see at this time.

Now everything works like I initially wanted.

Sorry if I sounded rude, I was just surprised that this wasn't included by default with the `Drawer`. I also wanted to let others know what to do in case they encounter the same situation as mine.

Thanks again! :)
